src/v/cluster/archival_metadata_stm.cc

```diff
@@ -350,18 +461,11 @@ ss::future<> archival_metadata_stm::handle_eviction() {
     }
 
     *_manifest = std::move(manifest);
-    for (const auto& segment : *_manifest) {
-        if (
-          _start_offset == model::offset{}
-          || segment.second.base_offset < _start_offset) {
-            _start_offset = segment.second.base_offset;
-        }
-    }
-    _last_offset = _manifest->get_last_offset();
+    auto start_offset = get_start_offset();
 
-    // We can skip all offsets up to the _last_offset because we can be sure
+    // We can skip all offsets up to the last_offset because we can be sure
     // that in the skipped batches there won't be any new remote segments.
-    _insync_offset = _last_offset;
+    _insync_offset = get_last_offset();
     auto next_offset = std::max(
       _raft->start_offset(), model::next_offset(_insync_offset));
     set_next(next_offset);
```

Comment:
Now that adding a segment is non-idempotent operation, this piece of code starts looking suspicious to me. Here we set the next offset of the STM somewhat arbitrarily, in the hope that replaying the commands that are already taken into account by the manifest state won't hurt, and skipping to `_last_offset + 1` is just an optimization. But now it is not true.

Maybe now that the STM manifest is the main one, we should save the offset of the last applied command in the manifest itself and use it here, I think this would help.

Reply:
implemented that

Reply:
in extra commit